### PR TITLE
Fix model loading hang

### PR DESCRIPTION
**Pull request by Tabrizian**

### CODE REVIEW

**Review by rmccorm4 — DISMISSED** [review]
What recently introduced this deadlock?

**Comment by Tabrizian** [comment]
@rmccorm4 Thanks for the quick review! Theoretically, this could have happened since we added support for thread pool in model loading. It could be that some changes in CUDA 12 have affected the timing and it has exposed this issue. There is a thread in the teams channel about why this could happen.

**Comment by rmccorm4** [comment]
> Theoretically, this could have happened since we added support for thread pool in model loading
Interesting. I believe even before the model-load thread pool change, new threads were spawned for each model load, so I think more or less the same level of concurrency existed before as well.

**Comment by Tabrizian** [comment]
I see, as long as we were using multiple threads for loading the models and removed persistent backends (i.e., backends that were always loaded if existed) this issue existed.

**Review by rmccorm4 — DISMISSED** [review]

**Comment by rmccorm4** [comment]
- Do we have a ticket for a more permanent fix? This seems more like a band-aid fix for now.

File: src/backend_manager.cc [hunk 1]
```diff
@@ -70,6 +74,16 @@ TritonBackend::Create(
...
   TritonServerMessage backend_config(backend_config_json);
 
+#ifdef TRITON_ENABLE_GPU
+  // This call is to prevent a deadlock issue with dlopening backend shared
+  // libraries and calling CUDA APIs in other threads. Since CUDA API also
+  // dlopens some libraries and dlopen has an internal lock, it can create a
+  // deadlock. Intentionally ignore the CUDA_ERROR (if any) for containers
+  // running on a CPU.
+  int device_count;
+  cudaGetDeviceCount(&device_count);
```
Comment (GuanLuo) [inline src/backend_manager.cc:84]: Should this be moved to shared library utils to make sure all shared libraries ops have CUDA init before accessing any libraries?

**Comment by Tabrizian** [comment]
> Do we have a ticket for a more permanent fix? This seems more like a band-aid fix for now.

There is not much we can do about this bug. It seems like a problem with dlopen/CUDA.

**Review by GuanLuo — APPROVED** [review]

**Review by rmccorm4 — APPROVED** [review]

**Pull request merged**